### PR TITLE
Fix rerunning a tool via API

### PR DESCRIPTION
**Pull request by nsoranzo**
When sending an API request `POST /api/tools` with payload:

```python
{
    'action': 'rerun',
    'history_id': history_id,
    'inputs': {},
    'target_dataset_id': target_dataset_id
    'tool_id': tool_id,
}
```

the following exception was raised:

```
Traceback (most recent call last):
  File "lib/galaxy/web/framework/decorators.py", line 283, in decorator
    rval = func(self, trans, *args, **kwargs)
  File "lib/galaxy/webapps/galaxy/api/tools.py", line 456, in create
    return self._create(trans, payload, **kwd)
  File "lib/galaxy/webapps/galaxy/api/tools.py", line 462, in _create
    return self._rerun_tool(trans, payload, **kwd)
  File "lib/galaxy/webapps/galaxy/api/tools.py", line 710, in _rerun_tool
    for action in tool.trackster_conf.actions:
AttributeError: 'NoneType' object has no attribute 'actions'
```

and after fixing this, another exception was raised:

```
Traceback (most recent call last):
  File "lib/galaxy/web/framework/decorators.py", line 283, in decorator
    rval = func(self, trans, *args, **kwargs)
  File "lib/galaxy/webapps/galaxy/api/tools.py", line 456, in create
    return self._create(trans, payload, **kwd)
  File "lib/galaxy/webapps/galaxy/api/tools.py", line 462, in _create
    return self._rerun_tool(trans, payload, **kwd)
  File "lib/galaxy/webapps/galaxy/api/tools.py", line 822, in _rerun_tool
    dataset_dict['track_config'] = self.get_new_track_config(trans, output_dataset)
  File "lib/galaxy/web/base/controller.py", line 1017, in get_new_track_config
    "filters": {'filters' : track_data_provider.get_filters()},
AttributeError: 'NoneType' object has no attribute 'get_filters'
```

Also add basic `get_filters()` method to `BaseDataProvider` class.

### CODE REVIEW

File: lib/galaxy/webapps/galaxy/api/tools.py [hunk 1]
```diff
@@ -625,13 +625,13 @@ def _rerun_tool(self, trans, payload, **kwargs):
         tool = trans.app.toolbox.get_tool(original_job.tool_id)
         if not tool or not tool.allow_user_access(trans.user):
             return trans.app.model.Dataset.conversion_messages.NO_TOOL
-        tool_params = dict([(p.name, p.value) for p in original_job.parameters])
+        tool_params = dict((p.name, p.value) for p in original_job.parameters)
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/api/tools.py]:
```suggestion
        tool_params = {p.name: p.value for p in original_job.parameters}
```

File: lib/galaxy/webapps/galaxy/api/tools.py [hunk 2]
```diff
@@ -625,13 +625,13 @@ def _rerun_tool(self, trans, payload, **kwargs):
...
 
         # TODO: rather than set new inputs using dict of json'ed value, unpack parameters and set using set_param_value below.
         # TODO: need to handle updates to conditional parameters; conditional
         # params are stored in dicts (and dicts within dicts).
         new_inputs = payload['inputs']
-        tool_params.update(dict([(key, dumps(value)) for key, value in new_inputs.items() if key in tool.inputs and value is not None]))
+        tool_params.update(dict((key, dumps(value)) for key, value in new_inputs.items() if key in tool.inputs and value is not None))
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/api/tools.py]:
```suggestion
        tool_params.update({key: dumps(value) for key, value in new_inputs.items() if key in tool.inputs and value is not None})
```

**Comment by mvdbeek** [comment]
Do we know when this was introduced ? Does this happen to all tools ? Should we add a test case if we don't have one ?

**Comment by nsoranzo** [comment]
@mvdbeek I think this API call was not used except for Trackster, and it may be still working when used from Trackster.

I was surprised that we don't have a tool rerun method in BioBlend and started to look at the Galaxy API. This is all we have at the moment, which I think it's missing some of the features available through the UI, but worked for the simplest case with this small modifications to BioBlend:

```diff
diff --git a/bioblend/galaxy/tools/__init__.py b/bioblend/galaxy/tools/__init__.py
index 4915593..2ccf22f 100644
--- a/bioblend/galaxy/tools/__init__.py
+++ b/bioblend/galaxy/tools/__init__.py
@@ -96,7 +96,7 @@ class ToolClient(Client):
         params['link_details'] = link_details
         return self._get(id=tool_id, params=params)
 
-    def run_tool(self, history_id, tool_id, tool_inputs):
+    def run_tool(self, history_id, tool_id, tool_inputs, **kwargs):
         """
         Runs tool specified by ``tool_id`` in history indicated
         by ``history_id`` with inputs from ``dict`` ``tool_inputs``.
@@ -119,6 +119,14 @@ class ToolClient(Client):
         payload = {}
         payload["history_id"] = history_id
         payload["tool_id"] = tool_id
+        action = kwargs.get('action')
+        if action:
+            payload['action'] = action
+            if action == 'rerun':
+                target_dataset_id = kwargs.get('target_dataset_id')
+                if not target_dataset_id:
+                    raise ValueError("target_dataset_id must be specified when action='rerun'")
+                payload['target_dataset_id'] = target_dataset_id
         try:
             payload["inputs"] = tool_inputs.to_dict()
         except AttributeError:
```

So, to answer your Qs:

> Do we know when this was introduced ? Does this happen to all tools ?

I think this never worked outside of Trackster.

> Should we add a test case if we don't have one ?

Yes, I can try to add one.

**Comment by mvdbeek** [comment]
> I was surprised that we don't have a tool rerun method in BioBlend and started to look at the Galaxy API. This is all we have at the moment, which I think it's missing some of the features available through the UI, but worked for the simplest case with this small modifications to BioBlend:

yeah, the UI first fetches the parameters by ~~posting~~ GET to api/job/<job_id>/build_for_rerun and then submits a new POST, that is probably preferable ?

> Yes, I can try to add one.

🥇

File: lib/galaxy/webapps/galaxy/api/tools.py (hunk 1 above)
Comment (nsoranzo) [reply lib/galaxy/webapps/galaxy/api/tools.py]: Fixed in a rebase, thanks!

**Comment by nsoranzo** [comment]
> > I was surprised that we don't have a tool rerun method in BioBlend and started to look at the Galaxy API. This is all we have at the moment, which I think it's missing some of the features available through the UI, but worked for the simplest case with this small modifications to BioBlend:
> 
> yeah, the UI first fetches the parameters by posting GET to api/job/<job_id>/build_for_rerun and then submits a new POST, that is probably preferable ?

Not sure what's the way to go here, maybe @jmchilton or @guerler have a better informed opinion.

> > Yes, I can try to add one.
> 
> :1st_place_medal:

Added.

**Review by mvdbeek — APPROVED** [review]

**Comment by jmchilton** [comment]
@jgoecks gave me permission to remove ``_rerun_tool`` over a year ago, I just haven't gotten around to it. I don't think it is an active path in the code and I'd prefer to invest in building up whatever API the UI uses.

**Comment by nsoranzo** [comment]
@jmchilton That's fine by me, the present API is anyway missing some features I'd like to see, namely having the option of replacing the failed datasets and collections in the history, and restarting jobs which were waiting on those.

Also, it returns a single dataset instead of the re-run job info.

**Comment by bgruening** [comment]
@nsoranzo @jmchilton should we get something similar in before we have the new API?

**Comment by nsoranzo** [comment]
I think there was a consensus to just have a new, better API endpoint, both @jmchilton and @mvdbeek offered to work on that :wink:

**Comment by bgruening** [comment]
Upps, this bit me again.

**Comment by mvdbeek** [comment]
This should be possible with https://github.com/galaxyproject/galaxy/pull/9802

**Pull request closed without merging**